scvi/autotune/_autotune.py

```diff
@@ -0,0 +1,346 @@
+        train
+            Whether to train the resulting best model. Defaults to `False`
+        **kwargs
+            Additional arguments for func:`ray.tune.run`
+
+        Returns
+        -------
+        A tuple with the best model object and tune Analysis object
+        """
+        if not scheduler:
+            scheduler = ASHAScheduler(
+                max_t=self.num_epochs, grace_period=1, reduction_factor=2
+            )
+
+        if not resources_per_trial:
+            if torch.cuda.is_available():
+                resources_per_trial = {"cpu": 1, "gpu": 1}
+            else:
+                resources_per_trial = {"cpu": 1}
+
+        if resources_per_trial["cpu"] > 1 and resources_per_trial["gpu"]:
+            resources_per_trial["gpu"] /= resources_per_trial["cpu"]
+
+        ray.init(num_cpus=resources_per_trial["cpu"], ignore_reinit_error=True)
+        analysis = tune.run(
+            self._trainable,
+            metric=metric,
+            mode=mode,
+            config=self.config,
+            num_samples=num_samples,
+            scheduler=scheduler,
+            progress_reporter=self.reporter,
+            name=name,
+            resources_per_trial=resources_per_trial,
+            local_dir=local_dir,
+            raise_on_failed_trial=False,
+            **kwargs,
+        )
+        ray.shutdown()
+        logger.info(
+            "Best hyperparameters found were: {}".format(analysis.get_best_config())
+        )
+        best_config = analysis.best_config
+        model_config = {}
+        trainer_config = {}
+        plan_config = {}
+        continious_covariate_config = {}
+        categorical_covariate_config = {}
+        hvg_config = {}
+        for key in best_config:
+            if key in self.model_hyperparams:
+                model_config[key] = best_config[key]
+            elif key in self.trainer_hyperparams:
+                trainer_config[key] = best_config[key]
+            elif key in self.plan_hyperparams:
+                plan_config[key] = best_config[key]
+            elif key in self.continious_covariates:
+                continious_covariate_config[key] = best_config[key]
+            elif key in self.categorical_covariates:
+                categorical_covariate_config[key] = best_config[key]
+            elif (
+                (self.test_effect_hvg and self.top_hvg) is not None
+                and key in self.test_effect_hvg
+                or self.top_hvg
+            ):
+                hvg_config[key] = best_config[key]
+
+        # retrieve and load best model
+        best_checkpoint = analysis.best_checkpoint
+
+        if "subset_adata" in hvg_config is True:
+            sc.pp.highly_variable_genes(
+                self.adata,
+                flavor="seurat_v3",
+                n_top_genes=hvg_config["top_n"],
+                batch_key=self.batch_key,
+                subset=True,
+            )
+
+        self.model.setup_anndata(
```

Comment:
to generalize to other models, we'll have to allow for passing in additional setup_args. They don't have to be tunable parameters, but maybe we could just add a `setup_args: Optional[Dict] = None` kwarg to the init fn